### ISSUE TITLE
[ENH] failing `polars` import with `DataFrame` not present

### ISSUE THREAD

**Issue by celestinoxp** [sktime/sktime#6651]
**Is your feature request related to a problem? Please describe.**
ImportError while loading conftest 'C:\Users\celes\pycaret\tests\conftest.py'.

```
tests\conftest.py:8: in <module>
    import pycaret.time_series.forecasting.functional
.venv\Lib\site-packages\pycaret\time_series\__init__.py:1: in <module>
    from pycaret.time_series.forecasting.functional import (
.venv\Lib\site-packages\pycaret\time_series\forecasting\functional.py:22: in <module>
    from pycaret.time_series.forecasting.oop import TSForecastingExperiment
.venv\Lib\site-packages\pycaret\time_series\forecasting\oop.py:18: in <module>
    from sktime.forecasting.base import BaseForecaster, ForecastingHorizon
.venv\Lib\site-packages\sktime\forecasting\base\__init__.py:9: in <module>
    from sktime.forecasting.base._base import BaseForecaster
.venv\Lib\site-packages\sktime\forecasting\base\_base.py:45: in <module>
    from sktime.base import BaseEstimator
.venv\Lib\site-packages\sktime\base\__init__.py:15: in <module>
    from sktime.base._base_panel import BasePanelMixin
.venv\Lib\site-packages\sktime\base\_base_panel.py:18: in <module>
    from sktime.datatypes import (
.venv\Lib\site-packages\sktime\datatypes\__init__.py:5: in <module>
    from sktime.datatypes._check import (
.venv\Lib\site-packages\sktime\datatypes\_check.py:32: in <module>
    from sktime.datatypes._hierarchical import check_dict_Hierarchical
.venv\Lib\site-packages\sktime\datatypes\_hierarchical\__init__.py:3: in <module>
    from sktime.datatypes._hierarchical._check import check_dict as check_dict_Hierarchical
.venv\Lib\site-packages\sktime\datatypes\_hierarchical\_check.py:49: in <module>
    from sktime.datatypes._panel._check import check_pdmultiindex_panel
.venv\Lib\site-packages\sktime\datatypes\_panel\__init__.py:3: in <module>
    from sktime.datatypes._panel._check import check_dict as check_dict_Panel
.venv\Lib\site-packages\sktime\datatypes\_panel\_check.py:50: in <module>
    from sktime.datatypes._series._check import (
.venv\Lib\site-packages\sktime\datatypes\_series\__init__.py:4: in <module>
    from sktime.datatypes._series._convert import convert_dict as convert_dict_Series
.venv\Lib\site-packages\sktime\datatypes\_series\_convert.py:41: in <module>
    from sktime.datatypes._registry import MTYPE_LIST_SERIES
.venv\Lib\site-packages\sktime\datatypes\_registry.py:61: in <module>
    from sktime.datatypes._table._registry import MTYPE_LIST_TABLE, MTYPE_REGISTER_TABLE
.venv\Lib\site-packages\sktime\datatypes\_table\__init__.py:4: in <module>
    from sktime.datatypes._table._convert import convert_dict as convert_dict_Table
.venv\Lib\site-packages\sktime\datatypes\_table\_convert.py:264: in <module>
    def convert_polars_eager_to_lazy(obj: pl.DataFrame, store=None) -> pl.LazyFrame:
E   AttributeError: module 'polars' has no attribute 'DataFrame'
```
**Describe the solution you'd like**
Maybe you should add polars to requirements or optional requiremets?

**Additional context**
I'm developing pycaret source-code to support scikit-learn 1.5 (https://github.com/pycaret/pycaret/pull/4009) and VsCode give me this error related to sktime and polars...
@fkiraly , Does it make sense for you to add polars to your sktime installation?

**Comment by fkiraly** [comment]
hm, this is strange. The `polars` import should happen only if it is present.

Can you share a `pip freeze` of the environment in question?

**Comment by celestinoxp** [comment]
More context: To avoid the error, I installed polars manually via "pip install polars" and the problem was resolved

im using windows 11 with python 3.11.9

pip freeze:

adagio==0.2.4
aiofiles==23.2.1
alembic==1.13.1
altair==5.3.0
aniso8601==9.0.1
annotated-types==0.7.0
ansi2html==1.9.1
antlr4-python3-runtime==4.11.1
anyio==4.4.0
appdirs==1.4.4
asttokens==2.4.1
attrs==23.2.0
beautifulsoup4==4.12.3
black==23.3.0
blinker==1.8.2
boto3==1.34.127
botocore==1.34.127
cachetools==5.3.3
catboost==1.2.5
category-encoders==2.6.3
certifi==2024.6.2
cffi==1.16.0
charset-normalizer==3.3.2
click==8.1.7
cloudpickle==3.0.0
cmdstanpy==1.2.3
colorama==0.4.6
colorlog==6.8.2
colour==0.1.5
comm==0.2.2
contourpy==1.2.1
cryptography==42.0.8
cycler==0.12.1
Cython==3.0.10
daal==2024.4.0
daal4py==2024.4.0
dacite==1.8.1
dash==2.17.1
dash-bootstrap-components==1.6.0
dash-core-components==2.0.0
dash-html-components==2.0.0
dash-table==5.0.0
dash-testing-stub==0.0.2
dash_auth==2.3.0
dash_cytoscape==1.0.1
dask==2024.6.0
dask-expr==1.1.3
debugpy==1.8.1
decorator==5.1.1
Deprecated==1.2.14
deprecation==2.1.0
dill==0.3.8
distributed==2024.6.0
distro==1.9.0
dnspython==2.6.1
docker==7.1.0
dtreeviz==2.2.2
dynaconf==3.2.5
email_validator==2.1.2
entrypoints==0.4
evidently==0.4.27
executing==2.0.1
explainerdashboard==0.4.7
fairlearn==0.7.0
Faker==25.8.0
fastapi==0.111.0
fastapi-cli==0.0.4
fastjsonschema==2.20.0
ffmpy==0.3.2
filelock==3.15.1
flake8==7.1.0
Flask==2.3.3
flask-simplelogin==0.1.2
Flask-WTF==1.2.1
fonttools==4.53.0
fs==2.4.16
fsspec==2024.6.0
fugue==0.8.7
fugue-sql-antlr==0.2.0
future==1.0.0
gevent==24.2.1
gitdb==4.0.11
GitPython==3.1.43
gradio==4.36.1
gradio_client==1.0.1
graphene==3.3
graphql-core==3.2.3
graphql-relay==3.2.0
graphviz==0.20.3
greenlet==3.0.3
gunicorn==22.0.0
h11==0.14.0
holidays==0.50
htmlmin==0.1.12
httpcore==1.0.5
httptools==0.6.1
httpx==0.27.0
huggingface-hub==0.23.4
hyperopt==0.2.7
idna==3.7
ImageHash==4.3.1
imbalanced-learn==0.12.3
importlib_metadata==7.1.0
importlib_resources==6.4.0
iniconfig==2.0.0
interpret==0.6.1
interpret-core==0.6.1
ipykernel==6.29.4
ipython==8.25.0
ipywidgets==8.1.3
isort==5.13.2
iterative-telemetry==0.0.8
itsdangerous==2.2.0
jedi==0.19.1
Jinja2==3.1.4
jmespath==1.0.1
joblib==1.3.2
jsonschema==4.22.0
jsonschema-specifications==2023.12.1
jupyter-dash==0.4.2
jupyter_client==8.6.2
jupyter_core==5.7.2
jupyterlab_widgets==3.0.11
kaleido==0.2.1
kiwisolver==1.4.5
kmodes==0.12.2
lightgbm==4.4.0
litestar==2.9.0
llvmlite==0.43.0
locket==1.0.0
lxml==5.2.2
m2cgen==0.10.0
Mako==1.3.5
Markdown==3.6
markdown-it-py==3.0.0
MarkupSafe==2.1.5
matplotlib==3.7.5
matplotlib-inline==0.1.7
mccabe==0.7.0
mdurl==0.1.2
mlflow==2.13.2
mlxtend==0.23.1
moto==4.2.14
msgpack==1.0.8
msgspec==0.18.6
multidict==6.0.5
multimethod==1.11.2
multiprocess==0.70.16
mypy-extensions==1.0.0
nbformat==5.10.4
nest-asyncio==1.6.0
networkx==3.3
nltk==3.8.1
numba==0.60.0
numpy==1.26.4
opentelemetry-api==1.25.0
opentelemetry-sdk==1.25.0
opentelemetry-semantic-conventions==0.46b0
optuna==3.6.1
optuna-integration==3.6.0
orjson==3.10.5
outcome==1.3.0.post0
oyaml==1.0
packaging==24.1
pandas==2.1.4
parso==0.8.4
partd==1.4.2
pathspec==0.12.1
patsy==0.5.6
percy==2.0.2
pexpect==4.9.0
phik==0.12.4
pillow==10.3.0
platformdirs==4.2.2
plotly==5.22.0
plotly-resampler==0.10.0
pluggy==1.5.0
pmdarima==2.0.4
polars==0.20.31
polyfactory==2.16.0
prometheus_client==0.20.0
prompt_toolkit==3.0.47
prophet==1.1.5
protobuf==4.25.3
psutil==5.9.8
ptyprocess==0.7.0
pure-eval==0.2.2
py4j==0.10.9.7
pyaml==24.4.0
pyarrow==15.0.2
pycaret @ file:///C:/Users/celes/pycaret
pycodestyle==2.12.0
pycparser==2.22
pydantic==2.7.4
pydantic_core==2.18.4
pydub==0.25.1
pyflakes==3.2.0
Pygments==2.18.0
pynndescent==0.5.12
pyod==2.0.0
pyparsing==3.1.2
pytest==8.2.2
python-dateutil==2.9.0.post0
python-dotenv==1.0.1
python-multipart==0.0.9
pytz==2024.1
PyWavelets==1.6.0
pywin32==306
PyYAML==6.0.1
pyzmq==26.0.3
qpd==0.4.4
querystring-parser==1.2.4
referencing==0.35.1
regex==2024.5.15
requests==2.32.3
responses==0.25.3
retrying==1.3.4
rich==13.7.1
rich-click==1.8.3
rpds-py==0.18.1
ruff==0.4.9
s3transfer==0.10.1
SALib==1.5.0
schemdraw==0.15
scikit-base==0.7.8
scikit-learn==1.5.0
scikit-learn-intelex==2024.4.0
scikit-optimize==0.10.2
scikit-plot==0.3.7
scipy==1.11.4
seaborn==0.13.2
selenium==3.141.0
semantic-version==2.10.0
shap==0.44.1
shellingham==1.5.4
six==1.16.0
sktime==0.30.1
slicer==0.0.7
smmap==5.0.1
sniffio==1.3.1
sortedcontainers==2.4.0
soupsieve==2.5
SQLAlchemy==2.0.30
sqlglot==25.1.0
sqlparse==0.5.0
stack-data==0.6.3
stanio==0.5.0
starlette==0.37.2
statsforecast==1.5.0
statsmodels==0.14.2
tbats==1.1.3
tbb==2021.12.0
tblib==3.0.0
tenacity==8.3.0
threadpoolctl==3.5.0
tomlkit==0.12.0
toolz==0.12.1
tornado==6.4.1
tqdm==4.66.4
traitlets==5.14.3
triad==0.9.7
trio==0.24.0
tsdownsample==0.1.3
typeguard==4.3.0
typer==0.12.3
typing-inspect==0.9.0
typing_extensions==4.12.2
tzdata==2024.1
ujson==5.10.0
umap-learn==0.5.6
urllib3==2.2.1
uvicorn==0.30.1
visions==0.7.6
waitress==3.0.0
watchdog==4.0.1
watchfiles==0.22.0
wcwidth==0.2.13
websockets==11.0.3
Werkzeug==2.3.8
widgetsnbextension==4.0.11
wordcloud==1.9.3
wrapt==1.16.0
WTForms==3.1.2
xgboost==2.0.3
xmltodict==0.13.0
xxhash==3.4.1
ydata-profiling==4.8.3
yellowbrick==1.5
zict==3.0.0
zipp==3.19.2
zope.event==5.0
zope.interface==6.4.post2

**Comment by fkiraly** [comment]
Hm, this is very strange.

The code must have passed the line `if _check_soft_dependencies(["polars", "pyarrow"], severity="none"):`

So, if it did, then `pl.DataFrame` should exist, because `polars` is in the environment.

Could you help diagnose and run this in your environment?

```
from sktime.utils.dependencies import _check_soft_dependencies

print(_check_soft_dependencies(["polars", "pyarrow"], severity="none"))
```

**Comment by celestinoxp** [comment]
```
from sktime.utils.dependencies import _check_soft_dependencies
print(_check_soft_dependencies(["polars", "pyarrow"], severity="none"))
```
True

**Comment by celestinoxp** [comment]
Now i uninstalled polars and have errors again...

![image](https://github.com/sktime/sktime/assets/41735311/ed3c0707-5854-4a69-91f3-5c42c3ba4473)

**Comment by celestinoxp** [comment]
@fkiraly something is wrong... without polars installed give True.... 

![image](https://github.com/sktime/sktime/assets/41735311/832060f1-6d4d-48b5-8830-3b583c78837e)

1 - maybe my VsCode have some trouble with configuration?
2- can you do the same in your machine and give me feedback?

**Comment by fkiraly** [comment]
that's very peculiar.

It could be you are running the code from a folder where `polars` is a valid object or variable name.

Can you check what `dir(polars)` or `import polars` (and then `dir(polars)`) gives you?

**Comment by celestinoxp** [comment]
> that's very peculiar.
> 
> It could be you are running the code from a folder where `polars` is a valid object or variable name (but not the widely known package).
> 
> Can you check what `dir(polars)` or `import polars` (and then `dir(polars)`) gives you?

after install polars again, then executing python in terminal:
>>> import polars
>>> dir(polars)

gives:
['Array', 'ArrowError', 'Binary', 'Boolean', 'Categorical', 'CategoricalRemappingWarning', 'ChronoFormatWarning', 'ColumnNotFoundError', 'ComputeError', 'Config', 'DATETIME_DTYPES', 'DURATION_DTYPES', 'DataFrame', 'DataType', 'Date', 'Datetime', 'Decimal', 'DuplicateError', 'Duration', 'Enum', 'Expr', 'FLOAT_DTYPES', 'Field', 'Float32', 'Float64', 'INTEGER_DTYPES', 'InProcessQuery', 'Int16', 'Int32', 'Int64', 'Int8', 'InvalidOperationError', 'LazyFrame', 'List', 'MapWithoutReturnDtypeWarning', 'NESTED_DTYPES', 'NUMERIC_DTYPES', 'NoDataError', 'Null', 'Object', 'OutOfBoundsError', 'PolarsDataType', 'PolarsError', 'PolarsPanicError', 'PolarsWarning', 'SQLContext', 'SchemaError', 'SchemaFieldNotFoundError', 'Series', 'ShapeError', 'String', 'StringCache', 'Struct', 'StructFieldNotFoundError', 'TEMPORAL_DTYPES', 'Time', 'UInt16', 'UInt32', 'UInt64', 'UInt8', 'Unknown', 'UnstableWarning', 'Utf8', '__all__', '__annotations__', '__builtins__', '__cached__', '__doc__', '__file__', '__loader__', '__name__', '__package__', '__path__', '__register_startup_deps', '__spec__', '__version__', '_cpu_check', '_reexport', '_utils', 'align_frames', 'all', 'all_horizontal', 'any', 'any_horizontal', 'api', 'apply', 'approx_n_unique', 'arange', 'arctan2', 'arctan2d', 'arg_sort_by', 'arg_where', 'build_info', 'business_day_count', 'coalesce', 'col', 'collect_all', 'collect_all_async', 'concat', 'concat_list', 'concat_str', 'config', 'contextlib', 'convert', 'corr', 'count', 'cov', 'cum_count', 'cum_fold', 'cum_reduce', 'cum_sum', 'cum_sum_horizontal', 'cumfold', 'cumreduce', 'cumsum', 'cumsum_horizontal', 'dataframe', 'datatypes', 'date', 'date_range', 'date_ranges', 'datetime', 'datetime_range', 'datetime_ranges', 'dependencies', 'disable_string_cache', 'duration', 'element', 'enable_string_cache', 'exceptions', 'exclude', 'expr', 'field', 'first', 'fold', 'format', 'from_arrow', 'from_dataframe', 'from_dict', 'from_dicts', 'from_epoch', 'from_numpy', 'from_pandas', 'from_records', 'from_repr', 'functions', 'get_index_type', 'groups', 'head', 'implode', 'int_range', 'int_ranges', 'io', 'last', 'lazyframe', 'len', 'lit', 'map', 'map_batches', 'map_groups', 'max', 'max_horizontal', 'mean', 'mean_horizontal', 'median', 'meta', 'min', 'min_horizontal', 'n_unique', 'nth', 'ones', 'os', 'plugins', 'polars', 'quantile', 'read_avro', 'read_clipboard', 'read_csv', 'read_csv_batched', 'read_database', 'read_database_uri', 'read_delta', 'read_excel', 'read_ipc', 'read_ipc_schema', 'read_ipc_stream', 'read_json', 'read_ndjson', 'read_ods', 'read_parquet', 'read_parquet_schema', 'reduce', 'repeat', 'rolling_corr', 'rolling_cov', 'scan_csv', 'scan_delta', 'scan_iceberg', 'scan_ipc', 'scan_ndjson', 'scan_parquet', 'scan_pyarrow_dataset', 'select', 'selectors', 'series', 'set_random_seed', 'show_versions', 'slice', 'sql', 'sql_expr', 'std', 'string_cache', 'struct', 'sum', 'sum_horizontal', 'tail', 'thread_pool_size', 'threadpool_size', 'time', 'time_range', 'time_ranges', 'type_aliases', 'using_string_cache', 'var', 'when', 'wrap_df', 'wrap_s', 'zeros']

**Comment by yarnabrina** [comment]
I will suggest to test in a new environment. If you create a new environment, install pycaret and your other dependencies as you want and do not install polars, do you get the same failure?

If you do not, which is what we expect, then I'll try to see if the site-packages of your current environment (or any other folder present in your `PYTHONPATH`) has any empty folder named `polars` or not.

**Comment by celestinoxp** [comment]
I think it's resolved!
In a deeper analysis I discovered that the reason may have been the anaconda update in which I updated python 3.10 to python 3.11, it turns out that my work in development folder possibly had some configuration from python 3.10....
It's a little strange when we don't know the real reason for the errors and the error messages aren't 100% clear either.

So I'm going to close this issue... and thank to @fkiraly and @yarnabrina. Thanks
